### DISCUSSION TITLE
not getting udp multicast working in the new mirrored network mode of WSL2, how to?

### DISCUSSION THREAD

**Discussion by LuukBerkel** [microsoft/WSL#10614] [Q&A] [answered]
According to the release information multicast should be supported in the new mirrored network mode of wsl2. Only when i try to use it i can only send multicast udp frames but i cannot receive them. Is this an issues with the windows firewall or does anyone else know, i have not enabled the new firewall option though?

This because i am trying to run ros2 in wsl2. And when i run ros2 multicast receive in wsl2 i am not able to receive any messages from an host on my local network. I and i am sure it is reachable because i can ping it and ssh into it. So when i run ros multicast send on a host on the same network, the ros2 multicast receive in my wsl2 should print hello world, but it doesn't. It does work the other way around though. 

Did anyone else get this working, else i will open an issue in the issues queue? I have also use sockperf to test but it also doesn't work, although i may have done it wrong. 

##Build: 
- WSL-versie: 2.0.4.0
- Kernelversie: 5.15.123.1-1
- Windowsversie: 22H2 on preview channel

**Comment by checkraisefold** [comment]
Can repo, WSL 2.0.4.0 with same of rest of version info.

**Comment by sacefe** [comment]
I have same problem described by Luuk using the latest prerelease 2.0.6    :(

**Reply by guitarpicva** [reply]
Same in WSL v 2.0.9

**Comment by checkraisefold** [comment]
Currently I was able to get multicast (through lan 255 address at least) working by turning on `hostAddressLoopback`. If it doesn't work, which sometimes it does break, rebooting my PC entirely (not restarting WSL) seemed to fix it. Could work for some people.

**Reply by sacefe** [reply]
Hi,  could you please explain more about the 255 address?

**Comment by kistlin** [comment]
For me it worked after adding a firewall exception. The rule was an Inbound rule to allow a certain UDP port range.

Except I have
[wsl2]
firewall=true

WSL version: 2.0.7.0
Kernel version: 5.15.133.1-1
WSLg version: 1.0.59
Windows version: 10.0.22631.2506

So maybe it's something with the firewall not working yet?

**Reply by uxduck** [reply]
Wouldn't turning on firewall for WSL traffic make things worse? I just noticed this value is on by default, so i'll try turn it off.

**Comment by borjamunozf** [comment]
I think multicast does not work with mirrored mode. 
- Tested with CONFIG_IP_MULTICAST=y kernel config enabled but same result.
- Tested with 2.1.15, 2.2.24 and 2.3.31
Applied multiple changes and rules in HyperV and Windows Firewall without success.
Tested with this example #12122 and the only way to make it work was setting up Bridged mode and external VMswitch.

My guess right now is that the internal VMSwitch that WSL uses by default does not allow multicast traffic or it's not able to _forward_ properly (ex: [hyperv multicast routing iss](https://answers.microsoft.com/en-us/windowserver/forum/all/need-help-configuring-multicast-routing-on-hyper-v/b1360719-0efb-4196-a403-54111e111e4b)

Also it works with WSL distro version **1**

If anyone is able to make the example I mentioned with Mirrored mode, please, let me know

**Reply by uxduck** [reply]
A relevant issue: https://github.com/microsoft/WSL/issues/10735

I ran wireshark **inside** WSL, and I can see multicast traffic being received. I run wireshark with `sudo wireshark` inside WSL though.

**Comment by sbrl** [comment]
According to <https://learn.microsoft.com/en-us/windows/wsl/networking#mirrored-mode-networking>, Windows 11 22H2 or higher is required to use mirrored mode networking?

**Comment by shixudong2020** [comment]
from local network->WSL2(mirrored),need CONFIG_IP_MULTICAST=y kernel config enabled or Run a multicast receiver simultaneously on the Windows host
multicast between Windows host and WSL2 (mirrored) not need CONFIG_IP_MULTICAST=y
1.Windows host-> WSL2 (mirrored)
Run a multicast receiver simultaneously on the Windows host. Then, when the Windows host sends a multicast packet, it also sends a copy to the loopback device, and mirrors the multicast packet to the receiving direction of the WSL2's mirrored NIC. In this way, WSL2 can receive the multicast packet from the Windows host smoothly.
2.WSL2 (mirrored)->Windows host
hostAddressLoopback=true,and WSL executes "sudo ip route add 224.1.1.1 via 169.254.73.152 dev eth0 onlink", Here eth0 denotes the NIC on which the default route is located.Then Windows host can successfully receive multicast packets from WSL2 (mirrored)
for detail,please see WSL2 and multicast(https://blog.csdn.net/sxd2001/article/details/144649381)

**Reply by borjamunozf** [reply]
> from local network->WSL2(mirrored),need CONFIG_IP_MULTICAST=y kernel config enabled or Run a multicast receiver simultaneously on the Windows host multicast between Windows host and WSL2 (mirrored) not need CONFIG_IP_MULTICAST=y 1.Windows host-> WSL2 (mirrored) Run a multicast receiver simultaneously on the Windows host. Then, when the Windows host sends a multicast packet, it also sends a copy to the loopback device, and mirrors the multicast packet to the receiving direction of the WSL2's mirrored NIC. In this way, WSL2 can receive the multicast packet from the Windows host smoothly. 2.WSL2 (mirrored)->Windows host hostAddressLoopback=true,and WSL executes "sudo ip route add 224.1.1.1 via 169.254.73.152 dev eth0 onlink", Here eth0 denotes the NIC on which the default route is located.Then Windows host can successfully receive multicast packets from WSL2 (mirrored) for detail,please see WSL2 and multicast(https://blog.csdn.net/sxd2001/article/details/144649381)

It does not work with mirrored.

I have tried so many things:
- My kernel is already with CONFIG_IP_MULTICAST=y
- I have loopbackAddress enabled = true
- Enabling/disabling firewall

Could you show your screenshots using https://github.com/microsoft/WSL/issues/12122 Python examples for multicast test between Windows>WSL and WSL>Windows and I'll believe that it works :)

Sadly this is a major issue in the company I work with

**Reply by shixudong2020** [reply]
when windows host have another NIC(e.g. VirtualBox Host-Only Ethernet Adapter has same metric as the default routing NIC like https://github.com/microsoft/WSL/issues/12122),
For WSL>Windows multicast test,You need to add one of the following two additional steps to the Windows host:
1.Specify the host multicast receiving NIC as the default routing NIC
Windows host multicast receiver should modify
mreq = struct.pack('4sL', group, socket.INADDR_ANY)
to
mreq = struct.pack('4sL', group, socket.inet_aton(interface_ip))
here interface_ip refer to the ip of the host's default routing NIC
or
2.Reduce the metric of the Windows host's default routing NIC(because Some NIC, such as wintun, do not allow you to change the metric easily )

**Reply by DanielBryars** [reply]
I used Set-NetFirewallHyperVVMSetting -Name '{40E0AC32-46A5-438A-A0B2-2B479E8F2E90}' -DefaultInboundAction Allow and I was able to receive inbound multicast traffic into WSL2 when using the new mirrored networking mode.

**Comment by AnthonyPhan** [comment, marked as answer]
I was able to get multicast traffic working with my WSL2 instance (for ROS2 development) using the steps below. Here are my notes incase it helps others:

- In the latest WSL2 Kernel `CONFIG_IP_MULTICAST` is actually enabled by default. You can confirm this by running
  ```
  zcat /proc/config.gz | grep CONFIG_IP_MULTICAST
  ```
- Enable `Mirror` mode networking and disable firewall. Your `.wslconfig` file should look like this. This can also technically be configured by the WSL2 Setting GUI.
  ```
  [wsl2]
  networkingMode=Mirrored
  firewall=false
  ```
- Run the following command in PowerShell window with admin privileges to Configure Hyper-V firewall settings to allow inbound connections:
  ```
  Set-NetFirewallHyperVVMSetting -Name '{40E0AC32-46A5-438A-A0B2-2B479E8F2E90}' -DefaultInboundAction Allow
  ```


Using the setup above i'm able to get ROS messages from other devices on the physical network (not just the host) into the WSL2 VM.

**Reply by southpolaris** [reply]
Thanks for slove the same problem. BTW what is this {40E0AC32-46A5-438A-A0B2-2B479E8F2E90} strings meaning.

**Reply by AnthonyPhan** [reply]
I think it's the GUID for the VM used by WSL2. It seems to be referenced a lot in Microsoft's docs.

**Reply by jublebi** [reply]
Worked for me as well. Had to change the "AllowHostPolicyMerge" to False as otherwise the settings are overridden by PolicyStores with higher priority (See https://learn.microsoft.com/en-us/powershell/module/netsecurity/set-netfirewallhypervvmsetting?view=windowsserver2025-ps)

**Reply by shixudong2020** [reply]
> I was able to get multicast traffic working with my WSL2 instance (for ROS2 development) using the steps below. Here are my notes incase it helps others:
> 
> * In the latest WSL2 Kernel `CONFIG_IP_MULTICAST` is actually enabled by default. You can confirm this by running
>   ```
>   zcat /proc/config.gz | grep CONFIG_IP_MULTICAST
>   ```
> * Enable `Mirror` mode networking and disable firewall. Your `.wslconfig` file should look like this. This can also technically be configured by the WSL2 Setting GUI.
>   ```
>   [wsl2]
>   networkingMode=Mirrored
>   firewall=false
>   ```
> * Run the following command in PowerShell window with admin privileges to Configure Hyper-V firewall settings to allow inbound connections:
>   ```
>   Set-NetFirewallHyperVVMSetting -Name '{40E0AC32-46A5-438A-A0B2-2B479E8F2E90}' -DefaultInboundAction Allow
>   ```
> 
> Using the setup above i'm able to get ROS messages from other devices on the physical network (not just the host) into the WSL2 VM.

are u sure u can get ROS messages from win11 host into the WSL2(mirrored)?
